utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,567 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE OverloadedLists #-}
+
+module Vulkan.Utils.Reflect
+  ( ShaderStage(..)
+  , ShaderInfo(..)
+  , reflection
+  , reflection'
+  , makeShaderInfo
+  , makeDescriptorInfo
+  , makeInputInfo
+  ) where
+
+import           Vulkan.CStruct.Extends         ( SomeStruct(..) )
+import           Vulkan.Core10.DescriptorSet    ( DescriptorSetLayoutBinding(..)
+                                                , DescriptorSetLayoutCreateInfo(..)
+                                                )
+import           Vulkan.Core10.Enums.DescriptorType
+                                                ( DescriptorType(..) )
+import           Vulkan.Core10.Enums.Format     ( Format(..) )
+import           Vulkan.Core10.Enums.VertexInputRate
+                                                ( VertexInputRate(..) )
+import           Vulkan.Core10.Pipeline         ( PipelineShaderStageCreateInfo(..)
+                                                , PipelineVertexInputStateCreateInfo(..)
+                                                , ShaderStageFlagBits(..)
+                                                , VertexInputAttributeDescription(..)
+                                                , VertexInputBindingDescription(..)
+                                                )
+import           Vulkan.Core10.Shader           ( ShaderModule(..)
+                                                , ShaderModuleCreateInfo(..)
+                                                )
+import           Vulkan.NamedType               ( (:::) )
+import           Vulkan.Zero                    ( zero )
+
+import           Control.Monad.IO.Class         ( MonadIO
+                                                , liftIO
+                                                )
+import           Control.Monad.Trans.Except     ( ExceptT
+                                                , except
+                                                , runExceptT
+                                                , throwE
+                                                )
+import           Data.Aeson                     ( (.:)
+                                                , (.:?)
+                                                , (.=)
+                                                , FromJSON(..)
+                                                , ToJSON(..)
+                                                , Value(String)
+                                                , eitherDecodeStrict'
+                                                , object
+                                                , withObject
+                                                , withText
+                                                )
+import           Data.Aeson.Types               ( Parser
+                                                , parseFail
```

Comment:
`parseFail` in `Data.Aeson.Types` module was added in `aeson` package since [1.4.7.0](https://github.com/haskell/aeson/blob/d711df76b826942f4a9e791712512c6b19b8c1c8/changelog.md#1470) and it is implement by `fail` from `base` package `Control.Monad.Fail` module.

Should we add a version bump for `aeson` or replace the `parseFail` with `fail`?

Reply:
Stackage LTS snapshot 14.27 (the last one for GHC 8.6.5) is on 1.4.6 :thinking: 

I'd prefer using `parseFail` in my apps, but this is a library and having broad compatibility should take priority.
Let's start with with `Prelude.fail` and see if project tests succeed (I expect they would).